Silence CRD Design Proposal (#5485)
prometheus-operator/prometheus-operator: Documentation/proposals/202304-silences.md

# Silence CRD

* Owners:
  * [mcbenjemaa](https://github.com/mcbenjemaa)
* Related Tickets:
  * [#5452](https://github.com/prometheus-operator/prometheus-operator/issues/5452)
  * [#2398](https://github.com/prometheus-operator/prometheus-operator/issues/2398)
* Other docs:
  * n/a

This document describes the creation of `Silence` Custom Resource Definition that defines Alertmanager silences
configurations in the Kubernetes way.

> # Why

Reviewer: I think we are missing a few reasons why we think this is useful:
1 - Multi tenancy: Alertmanager cannot be provided in read-only mode but silences CR can be using kubernetes rbac 
2 - Moving the giantswarm silence operator to prometheus operator would provide better community support than what we can currently offer

Author: I added them in the goals section

Reviewer: I think also we could mention that this can offer a better user experience so they can use the same tools to manage their monitoring pipeline (kyverno for instance, cr validations) whereas if silences are created via jobs or any other mechanism lire direct access to alertmanager, you have to redo this tooling again

Author: I have added it, in the goals section.
Please take a look

Prometheus-operator doesn't have a way to automate the management of Alertmanager silences. Users have either been using internal scripts
or an additional operator that does the job e.g. [silence-operator](https://github.com/giantswarm/silence-operator).

> * Users using CI/CD jobs to manage silences have reported that this is cumbersome, since they need to either expose

Reviewer: I would emphasize that this is even harder for multi-cluster setups as having flux and the operator running would simplify that (no need to complex CI setup or a job running in every cluster)

Author: I agree

# Silence CRD

* Owners:
  * [mcbenjemaa](https://github.com/mcbenjemaa)
* Related Tickets:
  * [#5452](https://github.com/prometheus-operator/prometheus-operator/issues/5452)
  * [#2398](https://github.com/prometheus-operator/prometheus-operator/issues/2398)
* Other docs:
  * n/a

This document describes the creation of `Silence` Custom Resource Definition that defines Alertmanager silences
configurations in the Kubernetes way.

# Why

Prometheus-operator doesn't have a way to automate the management of Alertmanager silences. Users have either been using internal scripts
or an additional operator that does the job e.g. [silence-operator](https://github.com/giantswarm/silence-operator).

* Users using CI/CD jobs to manage silences have reported that this is cumbersome, since they need to either expose
  Alertmanager via an ingress or using a service account to connect to the cluster, and this is even more complicated

>   in a multi-cluster setup.

Author: > Another use case where this would be useful is when enabling AlertmanagerConfig custom resources to offer self-service notifications.

what does this mean?

>   in a multi-cluster setup.

Reviewer: In GS, our original use case is that we needed to deploy the same silences in multiple clusters from a single source of truth.

Deploying flux is definitely easier on multiple clusters than deploying some internal tooling to sync silences and this is all done from within the cluster so using the alertmanager is enough.

In a CI way (unless you use something like tekton for sure) you would need to have your CI aware about all the different alertmanager URIs, but also have it on every silence which can be expensive depending on the number of clusters you have. Also, if the ci cannot access an alertmanager for some reason, you would need to rerun the CI again

  to fully manage the life cycle of Silences Custom resources, Nevertheless, this comes with a drawback

>   as the team must be able to manage a new component in their stack.

Author: could you check!

Additionally, having a new component in the stack and keeping it maintained is not always ideal (said the folks at [Giant Swarm](https://giantswarm.io) the owners of [silence-operator](https://github.com/giantswarm/silence-operator)),
because that would require us to fully manage this component as of any managed app (monitoring, logging, alerting etc..).

> Having the Silence CRD as part of prometheus-operator provides users with a better experience: there is no need to install another operator or build a system on top of it to provide that functionality.

Reviewer 2: I can see that having 2 components on the stack is more complex than 1, but here we're moving one component by making the other slightly more complex as well, i.e. users still need to monitor Prometheus-operator and Alertmanager to understand if their Silence CRs are behaving as expected.

Of course, I'm saying this without any experience with the silence-operator 😅, it might as well be a super simple component and I'm here making a fuss :P

Could you tell us a bit more about the common failure modes that you've found while using it at Giantswarm?

Author: We didn't have complex failures or monitoring for this component, we just need to have an alert that triggers if  a certain Silence reconciliation failed.

Reviewer: As @mcbenjemaa said, we have only 2 possible failure scenarios in the silence-operator, either the silence cannot be reconcilied with alertmanager (which almost never happen because we have a small Silence CRD API) or the flux kustomization pulling the silences if failing. 

If this hits into prometheus-operator, the failure modes would roughly be the same in my opinion

Author: Exactly

> By Adding support for `Silence` CRD in the prometheus-operator, this will make users more flexible in terms of choosing the tool
> to deploy their silences and will free others from managing a standalone component within the stack.

Reviewer 2: ```suggestion
Adding support for `Silence` CRD in the Prometheus-operator will provide more flexibility to users in terms of choosing the tool to deploy their silences and will free users of Giantswarm's silence operator from managing a standalone component within the stack.
```

I'm not sure about this suggestion, to be honest 😅 

"Free others" sounds too vague, but being very direct that this is helping users of Giantswarm home-grown solution sounds like an attack, which is not my intention 😬

Author: Yeah,

Author: @theo

Reviewer: I think one of the idea here is that new users would be able to use this operator without having to think how to manage their silences and existing users would be able to use it but they could also use their existing mechanism if they wish

Author: So @QuentinBisson what do you suggest to keep here?

Reviewer: I think Arthur suggestions make sense but maybe we do not need to talk specifically about the silence operator:

Adding support for `Silence` CRD in the Prometheus-operator will provide more flexibility to users in terms of choosing the tool to deploy their silences and will free users from managing a standalone component within the stack. 

That would work for both the silence operator or CI jobs or other tools that creative people came up with

Author: this is done

## Pitfalls of the current solution

Using Alertmanager API Directly comes with drawbacks:

* Teams have to build an automation to add silences in a centralized manner
* There is no input validation, which can lead to an invalid silence configuration
* Teams need to manage expiration on their own

> * If a silence was deleted by accident, it will be permanent.

Reviewer 2: I might be talking non-sense here, but wouldn't it be possible to implement a [finalizer](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/#:~:text=Finalizers%20are%20namespaced%20keys%20that,resources%20the%20deleted%20object%20owned.) here? Making sure the silence is revoked from alertmanager before the resource gets deleted.

Author: No, this section explains the drawback of using the API directly. 
in the Silence CR approach, this will be fixed by the control loops that will solve the drift.

# Goals

> * Provide a way for users to manage Alertmanager silences with the Kubernetes Way.

Reviewer 2: ```suggestion
* Provide a way for users to manage Alertmanager silences with Custom Resources (CR).
```

Kubernetes way might sound vague and ambiguous.

> * Make it easier to manage Alertmanager silences via centralised repo with e.g. ArgoCD, Flux...

Reviewer 2: ```suggestion
* Enable management of Alertmanager silences via centralised repo with e.g. ArgoCD, Flux...
```

> * Provide a way to manage silence expiration with a silence CR.

Reviewer 2: ```suggestion
```

The first goal already covers this one, doesn't it?

Author: I think yes

## Audience

* Users who serve Prometheus as a service and want to have an interface in defining silences exposed to developers.

> * Users who want to manage silences the same way as for services running within the Kubernetes cluster
> * Users who want a supported Kubernetes way of silences outside the Kubernetes cluster

Reviewer 2: ```suggestion
* Users who want to manage silences the same way as for services running within the Kubernetes cluster.
```

"Kubernetes way" sounds vague and ambiguous. And it is the same audience mentioned above right?

> * Provide a way for users to manage Alertmanager silences with Custom Resources (CR).

Reviewer 3: (nit)
```suggestion
* Provide a way for users to manage Alertmanager silences with a Custom Resource Definition (CRD).
```

* Enable management of Alertmanager silences via centralised repo with e.g. ArgoCD, Flux...
* Multi tenancy: Silences CR can be using kubernetes rbac.

> * Moving the Giant Swarm silence operator to prometheus operator would provide better community support

Reviewer 3: and less fragmentation?

Author: Yes :)

## Audience

> * Users who serve Prometheus as a service and want to have an interface in defining silences exposed to developers.

Reviewer 3: (nit)
```suggestion
* Users who use the Prometheus operator to offer monitoring-as-a-service to developers.
```

* Users who want to manage silences the same way as for services running within the Kubernetes cluster.

> * Users who want a supported Kubernetes way of silences outside the Kubernetes cluster

Reviewer 3: not sure I understand this one

Author: I think this is kind of highlight the GitOps approach

Reviewer: Maybe rephase it users using GitOps instead?

# Non-Goals

> * This proposal do not cover automatic cleanup of expired Silence CRs

Reviewer 3: expired silences are eventually removed by Alertmanager

Author: yes, But the CRs will remain in the cluster.

Reviewer: Well removing them once they expire should be easy but if you use flux, you would need to have an automation in place to remove expired silences. Maybe CR removal can be disabled?

Author: In the first release, we will not cover automatic cleanup, maybe we implement it, if requested

* Refactoring of the other CRDs is not in scope for the first version

# How

> Creating a new cluster-scope Silence CRD that will act as an interface by adding silences via the Alertmanager API.

Reviewer 3: hmm I would have expected a namespaced CRD?

Author: Yes, sure.

A typical `Silence` resource could look like the following:

```yaml
apiVersion: monitoring.coreos.com/v1alpha1
kind: Silence
metadata:
  name: my-silence
  labels:
    test: value
spec:
  selector:
    app: test-alertmanager

>   expiresAt: "2023-06-01"

Reviewer 3: Don't we need a datetime here?

Author: I think we need a format supported by the Alertmanager API. 
So, we will accept a string a validate that's a date or datetime.

Reviewer: The date is more of a remnant of our internal implementation but adding time definitely makes sense

Author: I will check this

  matchers:

>     - name: namespace

Reviewer 3: we should use the existing `Matcher` type from the AlertmanagerConfig CRD.

Author: Yes, I will check it

* `matchers` field corresponds to the Alertmanager silence matchers each of which consists of:
  - `name` - name of tag on an alert to match
  - `value` - fixed string or expression to match against the value of the tag named by name above on an alert
  - `isRegex` - a boolean specifying whether to treat value as a regex (=~) or a fixed string (=)
  - `isEqual` - a boolean specifying whether to use equal signs (= or =~) or to negate the matcher (!= or !~)

> This example doesn't list all the fields that are offered by prometheus. The implementation of all the fields will be

Reviewer 2: ```suggestion
This example doesn't list all the fields that are offered by Alertmanager. The implementation of all the fields will be
```

> `app: test-alertmanager`.

Reviewer 3: I don't think that it's the correct model. This approach would allow any user to create silences on any Alertmanager in the cluster.
IMHO we need `.spec.silenceSelector` and `.spec.silenceNamespaceSelector` in the Alertmanager CRD.

Author: From my point of view, I dont think this is a good idea  `.spec.silenceSelector` and `.spec.silenceNamespaceSelector` in the Alertmanager CRD.

As silences are An API resource and it's not a config like `AlertmanagerConfig`.

Having the selector in the Silence CR, make it easier to select which Alertmanager to Use. 

What do you think?

Author: From a security point of view, I guess we can limit the users by RBAC who can Create/Update Silences.

Author: If you emphasize using this approach, then we need to use both selectors in Alertmanager CRD.
and remove this selector.

Reviewer: I also thinks this makes more sense because silence creators should not be aware of which alertmanager they want to target

Reviewer 4: This is an interesting problem! I agree with @simonpasquier but I also see the other side where it would be great if the user did not have to care about which AM they are targeting.
For the sake of discussion, if we were to only have `selector`, would it be feasible for the operator to only accept Silences in the same place as the `PrometheusRule` or would this be too limiting?

Author: We will just go with selectors in the alertmanager for now: 
```yaml
[...]
spec:
  silenceSelector: ...
  silenceNamespaceSelector: ...
```

> done in an iterative process and as such, the expectation is not for all of them to be implemented in the first version.

Reviewer 2: I might be wrong, but the missing fields are just "Created By" and "Comment", aren't they? Which doesn't seem to offer a big challenge to implement when compared to expiration handling.

Couldn't we just add all fields all at once?

Author: we can add them as annotation, or you prefer fields under the spec?

Author: Any thoughts here?

Also, to help selecting `Silence`, a new field will be added to the Alertmanager CRD:

```yaml
[...]
spec:
  silenceSelector: ...
  silenceNamespaceSelector: ...
```

# Alternatives

> * Use Alertmanager API, with the pitfalls described earlier
> * Use [silence-operator](https://github.com/giantswarm/silence-operator), which is suboptimal as it requires users to install two operators for the feature

Reviewer 2: ```suggestion
* Use Alertmanager API.
* Use [silence-operator](https://github.com/giantswarm/silence-operator).
```

# Silence CRD

* Owners:
  * [mcbenjemaa](https://github.com/mcbenjemaa)
* Related Tickets:
  * [#5452](https://github.com/prometheus-operator/prometheus-operator/issues/5452)
  * [#2398](https://github.com/prometheus-operator/prometheus-operator/issues/2398)
* Other docs:
  * n/a

This document describes the creation of `Silence` Custom Resource Definition that defines Alertmanager silences
configurations in the Kubernetes way.

# Why

prometheus-operator doesn't have a way to automate the management of Alertmanager silences. Many users have either been using some internal scripts
or an additional operator that does the job e.g. [silence-operator](https://github.com/giantswarm/silence-operator).

* Users who are using CI/CD jobs to manage silences have reported that this is really cumbersome.
* Users that use scripts are not in better shape obviously.
* Users that use a standalone operator that exposes a Silence CRD are in a better situation, some folks have reported that they are using GitOps
  to fully manage the life cycle of Silences Custom resources, Nevertheless, this comes with a drawback
  as the team must be able to manage a new component in their stack.

> Additionally, Having a new component in the stack and keeping it maintained is not always ideal (said the folks at [Giant Swarm](https://giantswarm.io) the owners of [silence-operator](https://github.com/giantswarm/silence-operator)).

Reviewer 5: ```suggestion
Additionally, having a new component in the stack and keeping it maintained is not always ideal (said the folks at [Giant Swarm](https://giantswarm.io) the owners of [silence-operator](https://github.com/giantswarm/silence-operator)).
```

Can you elaborate as to why it's not ideal?

Author: done.

> It's really better to have this feature as part of prometheus-operator and keep it maintained and available for the community.

Reviewer 5: ```suggestion
Having the Silence CRD as part of prometheus-operator provides users with a better experience: there is no need to install another operator or build a system on top of it to provide that functionality.
```

By Adding support for `Silence` CRD in the prometheus-operator, this will make users more flexible in terms of choosing the tool
to deploy their silences and will free others from managing a standalone component within the stack.

## Pitfalls of the current solution

Using Alertmanager API Directly comes with drawbacks:

* Teams have to build an automation to add silences in a centralized manner
* There is no input validation, which can lead to an invalid silence configuration

> * Teams need to manage expiration on their own

Author: The intention here is that the controller will handle the expiration, via the status subresource.

> * Teams need to manage expiration on their own

Reviewer 2: I'm not sure if I understand the problem here. With a CRD, users still need to set an expiration date, don't they?

> * Teams need to manage expiration on their own

Author: yes

> * Teams need to manage expiration on their own

Author: the Alertmanager itself will handle the expiration. 
but, the actual Silence resource will still be deployed in the cluster.

I thought we could implement a cleanup mechanism once the silence has expired.

which means, if a silence is expired it will be deleted from the cluster as well as from Alertmanager

* If a silence was deleted by accident, it will be permanent.

# Goals

* Provide a way for users to manage Alertmanager silences with the Kubernetes Way.
* Make it easier to manage Alertmanager silences via centralised repo with e.g. ArgoCD, Flux...
* Provide a way to manage silence expiration with a silence CR.

## Audience

* Users who serve Prometheus as a service and want to have an interface in defining silences exposed to developers.
* Users who want to manage silences the same way as for services running within the Kubernetes cluster
* Users who want a supported Kubernetes way of silences outside the Kubernetes cluster

# Non-Goals

> * This proposal doesn't aim to remove the CR after the expiration date has reached

Reviewer 6: No operator should ever remove CR. But do you mean that silences won't be expired/removed from Alertmanager ?

Author: I was highlighting that maybe some users request the feature of automatic cleanup after the silences have expired.
We will not support that for now...

Reviewer 6: Alright then maybe re-phrase this to something like

```suggestion
* This proposal do not cover automatic cleanup of expired Silence CRs
```

Author: this is done

* Refactoring of the other CRDs is not in scope for the first version

# How

Creating a new cluster-scope Silence CRD that will act as an interface by adding silences via the Alertmanager API.
Usage of Silences doesn't exclude the use of the other CRDs, they are not mutually exclusive.
`Silence` will allow the creation of any silences to Alertmanager, while the other CRDs provide sane defaults. This will allow

> for isolated testing of the new `Silence` CRD.

Reviewer 5: I guess this is coming from the ScrapeConfig proposal, this doesn't make a lot of sense in the context of the Silence CRD. There is no overlap with the other CRDs.

Author: Ah Yes

> A typical Silence resouce should looks like the following:

Reviewer 5: ```suggestion
A typical `Silence` resource could look like the following:
```
We shouldn't mandate what it should look like, the implementation will take care of that.

> * Users who want a supported Kubernetes way of silences outside the Kubernetes cluster

Reviewer 4: These two sound the same. AFAIC as @QuentinBisson mentioned one of the goals is for users who manage their clusters strictly through GitOps

```yaml
apiVersion: monitoring.coreos.com/v1alpha1
kind: Silence
metadata:
  name: my-silence
  labels:
    test: value
spec:
  selector:
    app: test-alertmanager
  validUntil: "2023-06-01"
  matchers:
    - name: namespace
      value: test-ns
      isRegex: false
```

The above resource will result in creating a silence in the Alertmanager with label selector,
`app: test-alertmanager`.

> * `selector` used to select which Kubernetes Service to use as an Alertmanager API.

Reviewer 5: Shouldn't this be on the Prometheus CRD, which would match a label on the Silence? Same as `PrometheusRule`.

Author: Yes, I have added both strategies

Author: I thought there might be multiple alertmanagers in same cluster, 
but I also have a section where I define: 

```
  silenceSelector: ...
  silenceNamespaceSelector: ...
```

in the Alertmanager CRD.

Reviewer 2: Could we decide on using the selector in only one CRD?

Author: Yeah, this is not a configuration

Author: it's an API, so we just use this selector

Reviewer 5: Oh, sorry, I meant the `Alertmanager` CRD in my first comment 😅 

Another question though, how should one go about using the `Silence` CRD with a remote Alertmanager cluster? Or is that explicitely not covered in the proposal?

Author: Not Covered

Reviewer 4: We should probably add this to the `Non-goals` for this proposal

Author: yes

> * Refactoring of the other CRDs is not in scope for the first version

Reviewer: We can add a read-only alertmanager here behing a flag where the silence operator would automaticall remove silences created by hand

> * `validUntil` to define the expiration of the silence.

Reviewer 5: Would it make sense to have silences that stay active so long as the resource exists? For instance, if `validUntil` is undefined, then the Silence stays active. If it is, the `Silence` expires at the date defined.

Author: Yeah, I'm guessing if we can add a default maximum days if the user has not defined the `validUntil `

Author: we can make the field `expiresAt` required so :)

Reviewer 6: The implementation could also keep the silence active when this field is not defined. I have no strong opinion on this.

Reviewer 5: Got it, so:
* If a silence has `expiredAt`, it'll just expire at the time defined and the CR will have to be manually removed
* If a silence doesn't have `expiredAt`, it'll be kept active as long as the CR exists and removed as soon as the CR is gone

I'm not sure how do-able the second point is but I think it'd make sense, especially with GitOps.

Is that what we aim for?

Author: I think so,

Author: However,  how do we handle silence overwrite?

Reviewer: Silence operator would update them. The silence id in alertmanager would be kept in the status

Reviewer 5: Great! Can you add that to the proposal? I guess in the How section.

Author: Added!

> * `validUntil` to define the expiration of the silence.

Reviewer 6: I would rather align this with the Silence field `ExpiresAt`.

* `matchers` field corresponds to the Alertmanager silence matchers each of which consists of:
  - `name` - name of tag on an alert to match
  - `value` - fixed string or expression to match against the value of the tag named by name above on an alert
  - `isRegex` - a boolean specifying whether to treat value as a regex (=~) or a fixed string (=)
  - `isEqual` - a boolean specifying whether to use equal signs (= or =~) or to negate the matcher (!= or !~)

This example doesn't list all the fields that are offered by prometheus. The implementation of all the fields will be
done in an iterative process and as such, the expectation is not for all of them to be implemented in the first version.

Also, to help selecting `Silence`, a new field will be added to the Alertmanager CRD:

```yaml
[...]
spec:
  silenceSelector: ...
  silenceNamespaceSelector: ...
```

# Alternatives

> * Use Alertmanager API, with the pitfalls described earlier

Reviewer 5: ```suggestion
* Use Alertmanager API, with the pitfalls described earlier
* Use [silence-operator](https://github.com/giantswarm/silence-operator), which is suboptimal as it requires users to install two operators for the feature
```

# Silence CRD

* Owners:
  * [mcbenjemaa](https://github.com/mcbenjemaa)
* Related Tickets:
  * [#5452](https://github.com/prometheus-operator/prometheus-operator/issues/5452)
  * [#2398](https://github.com/prometheus-operator/prometheus-operator/issues/2398)
* Other docs:
  * n/a

This document describes the creation of `Silence` Custom Resource Definition that defines Alertmanager silences
configurations in the Kubernetes way.

# Why

> Prometheus-operator doesn't have a way to automate the management of Alertmanager silences. Users have either been using internal scripts

Reviewer: ```suggestion
Prometheus-operator does not have provide a way to automate the management of Alertmanager silences and there is no current standard/best practice way to do that.
Moreover, setting an automation to manage silences has been reported as cumbersome by multiple users.
```

or an additional operator that does the job e.g. [silence-operator](https://github.com/giantswarm/silence-operator).

> 

Author: Added!

# Silence CRD

* Owners:
  * [mcbenjemaa](https://github.com/mcbenjemaa)
* Related Tickets:
  * [#5452](https://github.com/prometheus-operator/prometheus-operator/issues/5452)
  * [#2398](https://github.com/prometheus-operator/prometheus-operator/issues/2398)
* Other docs:
  * n/a

This document describes the creation of `Silence` Custom Resource Definition that defines Alertmanager silences
configurations in the Kubernetes way.

# Why

> Prometheus-operator does not have provide a way to automate the management of Alertmanager silences and there is no current standard/best practice way to do that.
> Moreover, setting an automation to manage silences has been reported as cumbersome by multiple users.
> or an additional operator that does the job e.g. [silence-operator](https://github.com/giantswarm/silence-operator).

Reviewer 2: ```suggestion
Prometheus-operator does not provide a way to automate the management of Alertmanager silences and there is no current standard/best practice way to do that.
Moreover, setting an automation to manage silences has been reported as cumbersome by multiple users.
or an additional operator that does the job e.g. [silence-operator](https://github.com/giantswarm/silence-operator).
```

* Users using CI/CD jobs to manage silences have reported that this is cumbersome, since they need to either expose
  Alertmanager via an ingress or using a service account to connect to the cluster, and this is even more complicated
  in a multi-cluster setup.

>   However, there's folks that uses a Kubernetes Job to adds Silences,
>   In both situation, a CI/CD pipeline is needed in order to automate the execution, this includes managing the secrets and/or deprecation of code
>   In the Kubernetes Job way, teams need to build and deploy a container and monitor its status, which is an additional effort.
>   CI/CD pipeline is not the best option in multi-cluster setups, teams need to manage k8s access, secrets, and configs.

Reviewer 2: ```suggestion
  However, there are folks that use a Kubernetes Job to add Silences,
  In both situations, a CI/CD pipeline is needed in order to automate the execution, this includes managing the secrets and/or deprecation of code
  Using Kubernetes Jobs, teams need to build and deploy a container and monitor its status, which is an additional effort.
  CI/CD pipeline is not the best option in multi-cluster setups, teams need to manage k8s access, secrets, and configs.
```

![CI/CD K8s Job Approach](../img/CICD-k8s-job.png "CI/CD K8s Job Approach")

* Users that use a standalone operator that implements a Silence CRD are in a better situation,

>   for example users have reported that they are using GitOps to fully manage the life cycle of Silences Custom resources,

Reviewer 2: ```suggestion
  for example, users have reported that they are using GitOps to fully manage the life cycle of Silences Custom resources,
```

  which in fact brings benefits to the whole team because having a history and revisions in Git as well as constancy to manage the silences in a multi-tenant environment.

Additionally, having a new component in the stack and keeping it maintained is not always ideal (said the folks at [Giant Swarm](https://giantswarm.io) the owners of [silence-operator](https://github.com/giantswarm/silence-operator)),

> because that requires to manage an extra component +CRDs on top of the Prometheus operator.

Reviewer 2: ```suggestion
because that requires managing an extra component +CRDs on top of the Prometheus operator.
```

Having the Silence CRD as part of prometheus-operator provides users with a better experience: there is no need to install another operator or build a system on top of it to provide that functionality.

Adding support for `Silence` CRD in the Prometheus-operator will provide more flexibility to users in terms of choosing the tool to deploy their silences and will free users from managing a standalone component within the stack.

## Pitfalls of the current solution

Using Alertmanager API Directly comes with drawbacks:

* Teams have to build an automation to add silences in a centralized manner

> * There is no input validation, which can lead to an invalid silence configuration

Reviewer 2: I might be wrong, but I think that Alertmanager's API denies POST requests with invalid input. (definition can be seen [here](https://github.com/prometheus/alertmanager/blob/9a8d1f976e12b325ec47b84987a78b7845738be6/api/v2/openapi.yaml#L327-L347))

I understand that even the Alertmanager CLI, `amtool`, has [input validation](https://github.com/prometheus/alertmanager/blob/9a8d1f976e12b325ec47b84987a78b7845738be6/cli/silence_add.go#L73-L86) when adding a silence.

Could you elaborate on what kind of input validation that is lacking today?

Author: But In CI, there's no input validation, that's the intention.

Author: Changed

* Teams need to manage expiration on their own

> * If a silence was deleted by accident, it will be permanent.

Reviewer 2: I believe that it isn't possible to delete silences in Alertmanager actually. All we can do is expire the silence, and we can recreate them afterward. (Recreating is just creating a new one with fields pre-defined)

![image](https://user-images.githubusercontent.com/24193764/234670560-82ecd0ad-7b56-4ac7-9f1c-2150e7b30509.png)

Unless the user is using Alertmanager as a stateless container. If Alermanager is deleted, not the silence, then everything that was stored in Alertmanager's DB is gone, including the silences.

Author: YEs, the controller will handle the re-creation.

# Silence CRD

* Owners:
  * [mcbenjemaa](https://github.com/mcbenjemaa)
* Related Tickets:
  * [#5452](https://github.com/prometheus-operator/prometheus-operator/issues/5452)
  * [#2398](https://github.com/prometheus-operator/prometheus-operator/issues/2398)
* Other docs:
  * n/a

This document describes the creation of `Silence` Custom Resource Definition that defines Alertmanager silences.

# Why

Prometheus-operator does not provide a way to automate the management of Alertmanager silences and there is no current standard/best practice way to do that.
Moreover, setting an automation to manage silences has been reported as cumbersome by multiple users.

> or an additional operator that does the job e.g. [silence-operator](https://github.com/giantswarm/silence-operator).

Reviewer 3: this part doesn't read properly with the previous line.

Author: let me rephrase it.

  for example, users have reported that they are using GitOps to fully manage the life cycle of Silences Custom resources,
  which in fact brings benefits to the whole team because having a history and revisions in Git as well as constancy to manage the silences in a multi-tenant environment.

Additionally, having a new component in the stack and keeping it maintained is not always ideal (said the folks at [Giant Swarm](https://giantswarm.io) the owners of [silence-operator](https://github.com/giantswarm/silence-operator)),
because that requires managing an extra component +CRDs on top of the Prometheus operator.
Moreover, this project doesn't have a lot of community support, and it's reached by small amount of users.
Having the Silence CRD as part of prometheus-operator provides users with a better experience: there is no need to install another operator or build a system on top of it to provide that functionality.

Adding support for `Silence` CRD in the Prometheus-operator will provide more flexibility to users in terms of choosing the tool to deploy their silences and will free users from managing a standalone component within the stack.

## Pitfalls of the current solution

> Using Alertmanager API Directly comes with drawbacks:

Author: sure.

* Teams have to build an automation to add silences in a centralized manner

> * There is no input validation in CI, which can lead to an invalid silence configuration

Reviewer 3: The API will validate the request though so I'm not sure that it's a strong point.

Author: This means if you're using CI, it doesn't support validation prior to sending it to the API.
pre-validation is a benefit in this case.

* Teams need to manage expiration on their own

> * If a silence was deleted by accident, it will be permanent.

Author: so we need to delete this point or rephrase it?

# Goals

* Provide a way for users to manage Alertmanager silences with a Custom Resource Definition (CRD).
* Enable management of Alertmanager silences via centralised repo with e.g. ArgoCD, Flux...

> * Multi tenancy: Silences CR can be using kubernetes rbac.

Reviewer 3: (suggestion)
```suggestion
* Access to the Silence custom resources can be controlled by Kubernetes RBAC authorization.
```

> * Moving the Giant Swarm silence operator to prometheus operator would provide better community support and less fragmentation.

Reviewer 5: ```suggestion
* Provide better community support and reduce fragmentation as Giant Swarm's silence operator are added to prometheus-operator
```

* Improve user experience when managing silences, so users can use the same tools to manage their monitoring pipeline (kyverno, cr validations).

## Audience

* Users who use the Prometheus operator to offer monitoring-as-a-service to developers.
* Users who want to manage silences the same way as for services running within the Kubernetes cluster.
* Users who want a supported GitOps way of silences outside the Kubernetes cluster

# Non-Goals

* This proposal do not cover automatic cleanup of expired Silence CRs
* Refactoring of the other CRDs is not in scope for the first version

> * Adding silences to a remote Alertmanager cluster

Reviewer 3: not sure to understand this one. Is it about creating silences on an Alertmanager cluster not managed by the operator?

Author: yes, we will not cover adding silences to a remote alertmanager, at least not in the same cluster.

Reviewer 7: It can actually be easily achieved with optional field in the CRD for alertmanager url/auth/etc.

If it's not provided operator would take default one from config

# How

> Creating a new namespaced Silence CRD that will act as an interface by adding silences via the Alertmanager API.

Reviewer 5: ```suggestion
We will create a new namespaced Silence CRD that will act as an interface by adding silences via the Alertmanager API.
```

A typical `Silence` resource could look like the following:

```yaml
apiVersion: monitoring.coreos.com/v1alpha1
kind: Silence
metadata:
  name: my-silence
  namespace: monitoring
  labels:
    test: value
spec:
  expiresAt: "2023-06-01 10:00"
  matchers:
    - matchType: "="
      name: namespace
      value: test-ns
```

> The above resource will result in creating a silence in the Alertmanager that has the appropriate selector.

Reviewer 8: Will Silence custom resources only be applied on creation or regularly "pushed" to AM's API?
Is there a reconciliation loop (i.e. if there's a stateless AM and the AM pod gets recreated)? What's the suggested interval for it?

Author: yeah, there's reconciliation loop

Author: this will defined in the design document later

> 

Reviewer 8: What happens if a Silence CR cannot create a silence in AM? Can you describe certain error classes (e.g. invalid configuration, AM unavailable, etc.) and how they will be handled?

Author: Yes, the status sub resource, will has the error messages, 
Again, this is will be defined in the design document

* `expiresAt` to define the expiration of the silence.
  - If a silence has `expiredAt` defined, it'll expire at the time defined and the CR will have to be manually removed.
  - If a silence doesn't have `expiredAt`, it'll be kept active as long as the CR exists and removed as soon as the CR is gone
* `matchers` field corresponds to the Alertmanager silence matchers each of which consists of:
  - `name` - name of tag on an alert to match
  - `value` - fixed string or expression to match against the value of the tag named by name above on an alert
  - `matchType` - Match operation available with AlertManager >= v0.22.0 and takes precedence over Regex (deprecated) if non-empty.
  - `regex` - Whether to match on equality (false) or regular-expression (true). Deprecated as of AlertManager >= v0.22.0 where a user should use MatchType instead.

This example doesn't list all the fields that are offered by Alertmanager. The implementation of all the fields will be

> done in an iterative process and as such, the expectation is not for all of them to be implemented in the first version.

Author: comment field in the silence?

> done in an iterative process and as such, the expectation is not for all of them to be implemented in the first version.

Reviewer 7: Can we add CRD schema to the proposal?
I've posted mine above, I think it would be more transparent for everyone

> done in an iterative process and as such, the expectation is not for all of them to be implemented in the first version.

Reviewer 5: I think we should leave the schema to the implementation. We can definitely put some example uses but they shouldn't mandate how the API will look like later.

Also, to help select `Silences`, a new field will be added to the Alertmanager CRD:

```yaml
[...]
spec:

>   silenceSelector: ...

Reviewer 8: If this is selecting silences by labels, is it better to call it silenceLabelSelector ?

  silenceNamespaceSelector: ...
```

> Additionally, the silence-controller will handle `Silences` overwrite.

Reviewer 5: ```suggestion
Additionally, the silence-controller will handle `Silences` overwrites.
```

> after successfully created a silence, the silence id will be kept in the status sub-resource:

Reviewer 5: ```suggestion
After having successfully created a silence, the silence id will be kept in the status sub-resource:
```

> After having successfully created a silence, the silence id will be kept in the status sub-resource:

Reviewer 3: we could have the same silence being selected by different Alertmanager resources so this approach will not work out of the box.

Author: but. we are going to add `silenceSelector` and `silenceNamespaceSelector` to the alertmanager CRD

```yaml
[...]
status:

>   silenceID: ...

Reviewer 8: should this also `CreatedAt` `UpdatedAt` fields populated with AMdata, so we know that the resource got added to AM at a certain point in time?

Author: Yes, this will be handled in the Design document.
